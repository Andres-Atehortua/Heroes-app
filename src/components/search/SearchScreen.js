import { useMemo } from 'react';
import { useHistory, useLocation } from 'react-router';
import useForm from '../../hooks/useForm';
import { getHeroByName } from '../../selectors/getHeroByName';
import HeroCard from '../heroes/HeroCard';

const SearchScreen = () => {
  const history = useHistory();

  const location = useLocation();

  const params = new URLSearchParams(location.search);
  const query = params.get('q');

  const { values, handleInputChange } = useForm({
    input: query || '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    history.push(`?q=${values.input}`);
  };

  const heroesFiltered = useMemo(() => getHeroByName(query), [query]);
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />
      <div className='row'>
        <div className='col-5'>
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSubmit} className='d-grid gap-2'>
            <input
              className='form-control'
              type='text'
              placeholder='Find your hero'
              onChange={handleInputChange}
              value={values.input}
              name='input'
              autoComplete='off'
            />
            <button className='btn btn-outline-info mt-3'>Search</button>
          </form>
        </div>
        <div className='col-7'>
          <h4>Results</h4>
          <hr />

          {!query && <div className='alert alert-info'>Seach a hero</div>}
          {query && !heroesFiltered.length && (
            <div className='alert alert-danger'>
              There is no hero called "{query}"
            </div>
          )}

          {heroesFiltered.map((hero) => (
            <HeroCard {...hero} key={hero.id} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
