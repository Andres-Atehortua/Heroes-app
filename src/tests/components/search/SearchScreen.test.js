import { mount } from 'enzyme';
import { MemoryRouter, Route, Router } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Testing componente <SearchScreen />', () => {
  test('should match snapshot with default values', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('button').text()).toBe('Search');
  });

  test('should show the query value in the input', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=Batman']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('Batman');
  });

  test('should show error if hero is not found', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Route path='/search' component={SearchScreen} />
      </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').exists()).toBeTruthy();
    expect(wrapper.find('.alert-danger').text()).toBe(
      'There is no hero called "batman123"'
    );
  });

  test('should call history.push', () => {
    const history = { push: jest.fn(), listen: jest.fn(), location: {} };

    const wrapper = mount(
      <Router history={history}>
        <MemoryRouter initialEntries={['/search?q=batman123']}>
          <Route path='/search' component={SearchScreen} />
        </MemoryRouter>
      </Router>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: 'input',
        value: 'batman',
      },
    });
    wrapper.find('form').prop('onSubmit')({ preventDefault: () => {} });

    // expect(history.push).toHaveBeenCalled();
  });
});
