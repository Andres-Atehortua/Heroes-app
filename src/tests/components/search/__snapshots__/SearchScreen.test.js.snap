// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing componente <SearchScreen /> should match snapshot with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="d-grid gap-2"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="input"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-info mt-3"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Seach a hero
      </div>
    </div>
  </div>
</div>
`;
